### DISCUSSION TITLE
Allowing instance attributes for models

### DISCUSSION THREAD

**Discussion by eudoxos** [pydantic/pydantic#2459] [Question] [answered]
I am using a RPC library which needs to create certain special instance attributes for its own internal use for objects which get shared over the network. Can I request a model to allow creating of instance attributes? I did not see anything in the documentation (besides [Private attributes](https://pydantic-docs.helpmanual.io/usage/models/#private-model-attributes), but those need to be declared beforehand).

Unfortunately, it is not enough to set `_someAttr1 to `None`, the lib assumes it does not exist at all.

Thanks!

```python
import pydantic
class A(pydantic.BaseModel): pass
a=A()
a._someAttr='value'
```
The result is:

```python
ValueError: "A" object has no field "_someAttr"
```

I am able to work around it as follows, but I am not sure if it does not mess up some other pydantic internals. Plus, obviously, it is not very elegant.

```python
import pydantic
class A(pydantic.BaseModel):
    pass
    def __setattr__(self,a,v):
        if a.startswith('_'): object.__setattr__(self,a,v)
        else: super().__setattr__(a,v)
a=A()
a._someAttr='value' # this is OK
a.somethingelse=3 # this fails
```

**Comment by samuelcolvin** [comment, marked as answer]
```py
from pydantic import BaseModel

class A(BaseModel):

    class Config:
        extra = 'allow'
```

and it should work.

**Reply by eudoxos** [reply]
So many things ready for me :) Thanks again.

**Reply by eudoxos** [reply]
Alright, after some more testing, this is also allowing abitrary keywords to the ctor. What I am looking for that I can explicitly assign instance variables, but pydantic will refuse unknown attributes in the constructor. The intent is that explicit assignment will pass, but regular user who only consumes the class interface will be prevented from making mistakes.

```python
import pydantic
class A(pydantic.BaseModel):
    class Config:
        extra='allow'
    def __init__(self,**kw):
        super().__init__(**kw)
        self.this_is_something_I_want=3 # good
a=A(garbage=3454) # but this passes as well
```

**Reply by samuelcolvin** [reply]
Best to do it yourself using `__setattr__`, this is not something pydantic supports.

**Reply by eudoxos** [reply]
I achieved that by writing a custom `__init__(*args,**kwargs)` in the base model which raises exception when there is a positional argument, or a kw which is not in `self.__class__.__fields__`. That works well, in conjunction with `extra='allow'`.

**Comment by samuelcolvin** [comment]
out of interest, what RPC library are you using?

**Reply by eudoxos** [reply]
[Pyro5](https://github.com/irmen/Pyro5/)
